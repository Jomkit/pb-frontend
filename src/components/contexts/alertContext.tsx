import { createContext } from "react";

const alertContext = createContext({} as {setAlertMessage: Function, setAlertOn: Function});

export default alertContext