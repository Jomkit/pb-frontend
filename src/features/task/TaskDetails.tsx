import { useLocation, useNavigate } from 'react-router-dom';
import Frame from '../../components/Frame';
import Card from '../../components/Card';
import Timer from '../timer/Timer';
import { ITask } from '../../types';
import Api from '../../api';
import { useState } from 'react';

const TaskDetails = () => {
  // const projectId = localStorage.get("currProjectId");
  const { state }: {state: ITask} = useLocation();
  const navigate = useNavigate();
  const [deleteBtnDisabled, setDeleteBtnDisabled] = useState(false);

  const handleDelete = async () => {
    setDeleteBtnDisabled(true);
    const deletedTask = await Api.deleteTask(state.projectId as string, state.id as string);

    console.log(deletedTask);
    alert(deletedTask);
    navigate("/tasks");
  }

  return (
    <Frame bgColor="bg-orange-400" title="Task Details">
      <Card title={state.name}> 
        <div>
          <button className='btn font-bold bg-red-400' onClick={handleDelete} disabled={deleteBtnDisabled}>Delete</button>
        </div>
          <div className='m-2'>
            <Timer projectId={state.projectId as string} task={state} />
          </div>
      </Card>
    </Frame>
  )
}

export default TaskDetails