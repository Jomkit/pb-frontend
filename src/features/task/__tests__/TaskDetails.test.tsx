import { render } from "@testing-library/react"
import TaskDetails from "../TaskDetails"
import { MemoryRouter } from "react-router-dom";
import alertContext from "../../../components/contexts/alertContext";
import userEvent from "@testing-library/user-event";
import Api from "../../../api";

const mockedSetAlertMessage = vi.fn();
const mockedSetAlertOn = vi.fn();

beforeEach(() => {
    vi.mock("../../../api");
    vi.mock("react-router-dom", async (importOriginal) => {
        const mod = await importOriginal<typeof import("react-router-dom")>();
        return {
            ...mod,
            useLocation: () => ({state: {id: "1", name: "testTask 1", projectId: "1"}}),
        }
    })
})
afterEach(() => {
    vi.restoreAllMocks();
})

it("should render without crashing", () => {
    render(
        <MemoryRouter>
            <alertContext.Provider value={{setAlertMessage: mockedSetAlertMessage, setAlertOn: mockedSetAlertOn}} >
                <TaskDetails />
            </alertContext.Provider>
        </MemoryRouter>
    );
})

it("should match snapshot", () => {
    const { asFragment } = render(
        <MemoryRouter>
            <alertContext.Provider value={{setAlertMessage: mockedSetAlertMessage, setAlertOn: mockedSetAlertOn}} >
                <TaskDetails />
            </alertContext.Provider>
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
})

it("should call Api.deleteTask when delete button pressed", async () => {
    const deleteTaskRef = vi.spyOn(Api, "deleteTask");
    const user = userEvent.setup();
    
    const { findByText } = render(
        <MemoryRouter>
            <alertContext.Provider value={{setAlertMessage: mockedSetAlertMessage, setAlertOn: mockedSetAlertOn}} >
                <TaskDetails />
            </alertContext.Provider>
        </MemoryRouter>
    );
    const deleteBtn = await findByText("Delete");
    expect(deleteBtn).toBeInTheDocument();

    await user.click(deleteBtn);
    expect(deleteTaskRef).toHaveBeenCalled();
    expect(deleteTaskRef).toHaveBeenCalledWith("1", "1");
})

it("should disable delete button after it is pressed", async () => {
    const deleteTaskRef = vi.spyOn(Api, "deleteTask");
    const user = userEvent.setup();
    
    const { findByText } = render(
        <MemoryRouter>
            <alertContext.Provider value={{setAlertMessage: mockedSetAlertMessage, setAlertOn: mockedSetAlertOn}} >
                <TaskDetails />
            </alertContext.Provider>
        </MemoryRouter>
    );
    const deleteBtn = await findByText("Delete");
    expect(deleteBtn).toBeInTheDocument();

    expect(deleteBtn).not.toBeDisabled();
    await user.tripleClick(deleteBtn);
    expect(deleteBtn).toBeDisabled();
    expect(deleteTaskRef).toHaveBeenCalled();
    expect(deleteTaskRef).toHaveBeenCalledWith("1", "1");
})