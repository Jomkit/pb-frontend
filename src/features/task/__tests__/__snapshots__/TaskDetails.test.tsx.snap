// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <section
    class="text-center items-start min-h-screen w-screen p-4 bg-orange-400"
  >
    <h1>
      Task Details
    </h1>
    <div
      class="p-10 justify-center"
    >
      <div
        class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
      >
        <div
          class="text-center"
        >
          <h2>
            testTask 1
          </h2>
          <h4 />
        </div>
        <div>
          <button
            class="btn font-bold bg-red-400"
          >
            Delete
          </button>
        </div>
        <div
          class="m-2"
        >
          <div
            class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
          >
            <div
              class="text-center"
            >
              <h2 />
              <h4 />
            </div>
            <h2
              class="text-start"
            >
              Timer
            </h2>
            <div
              class="flex justify-between"
            >
              <div
                class="text-center"
              >
                <h3
                  class="text-2xl bg-slate-200 w-fit m-auto px-4 py-1 rounded border-2 border-slate-500"
                >
                  25:00
                </h3>
              </div>
              <div
                id="set-timer"
              >
                <button
                  class="btn bg-green-400"
                  id="start-btn"
                >
                  <i
                    class="fa-solid fa-play"
                  />
                </button>
                <button
                  class="btn bg-red-500"
                  id="restart-btn"
                >
                  <i
                    class="fa-solid fa-rotate-right"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
